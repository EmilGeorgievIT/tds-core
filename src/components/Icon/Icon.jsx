import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const Icon = ({ glyph, variant, fixedWidth, className, children, ...rest }) => {
  const classes = classNames(
    'icon',
    `icon-core-${glyph}`,
    className,
    {
      'icon--fw': fixedWidth,
      [`icon--${variant}`]: variant
    }
  );

  return (
    <i {...rest} className={classes}>
      {children}
    </i>
  );
};

Icon.propTypes = {
  /**
   * Name of the icon glyph.
   */
  glyph: PropTypes.oneOf([
    'caret-down',
    'caret-up',
    'checkmark',
    'chevron',
    'exclamation-point-circle',
    'expander',
    'hamburger',
    'incomplete',
    'location',
    'minus',
    'plus',
    'question-mark-circle',
    'spyglass',
    'times'
  ]).isRequired,
  /**
   * The appearance of the Icon.
   */
  variant: PropTypes.oneOf([
    'primary',
    'secondary',
    'disabled',
    'error'
  ]),
  /**
   * Whether or not to give the icon a fixed width.
   */
  fixedWidth: PropTypes.bool,
  /**
   * One or more CSS class names separated by spaces to append onto the icon.
   * Don't advertise as we plan on removing this feature soon.
   *
   * @ignore
   */
  className: PropTypes.string,
  /**
   * This can be used to add screen-reader content into the icon, but it must
   * be hidden from view. This behaviour is better accomplished with aria-* attributes.
   *
   * @ignore
   */
  children: PropTypes.node
};
Icon.defaultProps = {
  variant: null,
  fixedWidth: false,
  className: '',
  children: null
};

export default Icon;
