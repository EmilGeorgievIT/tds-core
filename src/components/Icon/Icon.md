### Available Icons

```
<div className="docs--horizontal-spacing">
  <Icon glyph="caret-down" />
  <Icon glyph="caret-up" />
  <Icon glyph="checkmark" />
  <Icon glyph="left-chevron" />
  <Icon glyph="chevron" />
  <Icon glyph="exclamation-point-circle" />
  <Icon glyph="expander" />
  <Icon glyph="hamburger" />
  <Icon glyph="incomplete" />
  <Icon glyph="location" />
  <Icon glyph="minus" />
  <Icon glyph="plus" />
  <Icon glyph="question-mark-circle" />
  <Icon glyph="spyglass" />
  <Icon glyph="times" />
</div>
```

### Modifying color

Use the `variant` prop to alter the icon's color. Each variant has semantic meaning.

```
<div className="docs--horizontal-spacing">
  <Icon glyph="checkmark" variant="primary" />
  <Icon glyph="incomplete" variant="secondary" />
  <Icon glyph="exclamation-point-circle" variant="error" />
</div>
```

#### Primary and secondary

Indicates a primary or secondary action.

```
<div>
  <div><Icon glyph="plus" variant="primary" /> Add a user</div>
  <div><Icon glyph="minus" variant="secondary" /> Remove a user</div>
</div>
```

#### Error

Indicates a problem.

```
<div>
  <Icon glyph="location" variant="error" /> Your location needs to be updated
</div>
```

### Accessibility considerations

Icons can be either decorative or meaningful.

**Decorative icons** do not perform a role beyond visual aesthetics and should be hidden from screen readers using the
`aria-hidden` attribute. Usually, the information being communicated with the icon is also conveyed in another manner.

This example shows a decorative icon that is hidden from screen readers. The Icon component sets `aria-hidden` to `true` by default; you can inspect the element and see.

```
<Paragraph>
  <Icon glyph="location" /> You are located in British Columbia.
</Paragraph>
```

**Meaningful icons** have meaning within the context of the page, which should be communicated to screen readers with the
`aria-label` attribute. Meaningful icons can also be interactive elements, which should be designated with
the `role` prop.

This example shows a meaningful icon that needs accessibility attributes.

```
<Paragraph>
  <button style={{appearance: 'none', background: 'none', border: 0, cursor: 'pointer'}} type="submit"><Icon glyph="spyglass" label="search" /></button>
</Paragraph>
```
