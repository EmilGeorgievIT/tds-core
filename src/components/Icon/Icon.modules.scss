@import '../../scss/settings/icons';
@import '../../scss/settings/colours';
@import '../../scss/utility/icons-utility';

// Unicode points of each icon's location within the font files.
$core-icon-codepoints: (
  caretDown: "\f105",
  caretUp: "\f106",
  checkmark: "\f101",
  chevron: "\f107",
  leftChevron: "\f107",
  exclamationPointCircle: "\f103",
  expander: "\f113",
  hamburger: "\f112",
  incomplete: "\f104",
  location: "\f110",
  minus: "\f109",
  plus: "\f108",
  questionMarkCircle: "\f102",
  spyglass: "\f111",
  times: "\f104"
);

.icon {
  @include core-icon;

  transition: color 0.1s linear;
}

.primary {
  color: $color-icon-primary;
}

.secondary {
  color: $color-icon-secondary;
}

.inverted {
  color: $color-white;
}

.error {
  color: $color-cardinal;
}

.disabled {
  color: $color-icon-disabled;
}

.fw {
  width: 1.09rem;
  text-align: center;
}

.small {
  font-size: 1 rem;
}

.medium {
  font-size: 1.5rem;
}

.large {
  font-size: 3rem;
}

@each $name, $codepoint in $core-icon-codepoints {
  .iconCore#{to-upper-case($name)}::before {
    // composes: icon; TODO: can't use composes inside sass @each and
    // composes doesn't quite make sense on the color modifiers
    content: $codepoint;
  }
}

.iconCoreLeftChevron::before {
  display: inline-block;
  transform: rotate(-180deg) translateY(1.5px);
}
