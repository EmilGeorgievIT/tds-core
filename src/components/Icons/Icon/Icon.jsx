import React from 'react'
import PropTypes from 'prop-types'

import safeRest from '../../../safeRest'
import joinClassNames from '../../../utils/joinClassNames'
import capitalize from '../../../capitalize'

import styles from './Icon.modules.scss'

const Icon = ({ symbol, variant, size, ...rest }) => {
  const classes = joinClassNames(
    styles[`icon${capitalize(symbol)}`],
    variant && styles[variant],
    size && styles[`size${size}`]
  )

  return (
    <i {...safeRest(rest)} className={classes} />
  )
}

Icon.propTypes = {
  symbol: PropTypes.oneOf([
    'caretDown',
    'caretUp',
    'checkmark',
    'chevron',
    'leftChevron',
    'exclamationPointCircle',
    'expander',
    'hamburger',
    'incomplete',
    'location',
    'minus',
    'plus',
    'questionMarkCircle',
    'spyglass',
    'times'
  ]).isRequired,
  variant: PropTypes.oneOf([
    'primary',
    'secondary',
    'inverted',
    'error'
  ]),
  size: PropTypes.oneOf([16, 24, 48])
}

Icon.defaultProps = {
  variant: undefined,
  size: 24
}

export default Icon
