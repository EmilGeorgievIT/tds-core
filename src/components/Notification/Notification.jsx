import React from 'react'
import PropTypes from 'prop-types'

import safeRest from '../../utils/safeRest'

import Grid from '../../old-components/Grid/Grid'
import DecorativeIcon from '../Icons/DecorativeIcon/DecorativeIcon'
import ColoredTextProvider from '../Typography/ColoredTextProvider/ColoredTextProvider'
import Paragraph from '../Typography/Paragraph/Paragraph'
import Flexbox from '../Flexbox/Flexbox'
import Box from '../Box/Box'

import messagingStyles from '../Messaging.modules.scss'
import iconWrapperStyles from '../Icons/IconWrapper.modules.scss'
import styles from './Notification.modules.scss'

const {Container, Row, Column} = Grid

const iconByVariant = {
  success: {
    symbol: 'checkmark',
    color: 'primary',
  },
  error: {
    symbol: 'exclamationPointCircle',
    color: 'error',
  },
}

const isImportant = variant => variant === 'success' || variant === 'error'

const renderIcon = icon => (
  <Box right={3} dangerouslyAddClassName={iconWrapperStyles.fixLineHeight}>
    <DecorativeIcon symbol={icon.symbol} variant={icon.color} />
  </Box>
)

const renderContent = (variant, children) => {
  const content = <Paragraph bold={isImportant(variant)}>{children}</Paragraph>

  if (variant === 'error') {
    return (
      <ColoredTextProvider colorClassName={messagingStyles.errorText}>
        {content}
      </ColoredTextProvider>
    )
  }

  return content
}

/**
 * A banner that highlights important messages.
 */
const Notification = ({variant, children, ...rest}) => (
  <Box {...safeRest(rest)} vertical={3} dangerouslyAddClassName={styles[variant]}>
    <Container limitWidth>
      <Row>
        <Column xs={12}>
          <Flexbox direction="row">
            {isImportant(variant) ? renderIcon(iconByVariant[variant]) : undefined}

            {renderContent(variant, children)}
          </Flexbox>
        </Column>
      </Row>
    </Container>
  </Box>
)

Notification.propTypes = {
  /**
   * The appearance.
   */
  variant: PropTypes.oneOf(['instructional', 'branded', 'success', 'error']),
  /**
   * The message. Can be raw text or text components.
   */
  children: PropTypes.node.isRequired,
}

Notification.defaultProps = {
  variant: 'instructional',
}

export default Notification
