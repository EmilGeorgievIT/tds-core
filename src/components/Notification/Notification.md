## Minimal usage

```
<Notification>Login to modify your account preferences.</Notification>
```

## Variants

Use the `variant` prop to alter the Notification's appearance.


### Branded

```
<Notification variant="branded">Welcome to TELUS.</Notification>
```

### Success

```
<Notification variant="success">Thank you for registering.</Notification>
```

### Error

```
<Notification variant="error">
  <p>Looks like our registration system is temporarily down.</p>
  <p>Come back another time to register for My Account.</p>
</Notification>
```

### Instructional

```
<Notification variant="instructional">
  Fill out the form below, then click <em>"Next"</em> to continue.
</Notification>
```
