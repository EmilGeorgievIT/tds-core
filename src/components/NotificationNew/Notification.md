## Minimal usage

Used for all general or instructional type messages. Defaults to instructional.

Only pass in text. Write in complete sentences.

```
<Notification>
  <strong>Tip:</strong> The services are best suited for larger business organizations ordering more than 50 plans on one account.
</Notification>
```

## Variants

Use the `variant` prop to alter the Notification's appearance.


### Branded

Used for feedback or chat related messages.

```
<Notification variant="branded">
  <strong>Tell us what you think</strong> It’s in our nature to listen. As TELUS.com continues to evolve, we’d love to hear more from you.
</Notification>
```

### Success

Used for successful transactional messages. Text is bolded to show importance.

```
<Notification variant="success">
  Your password has been successfully changed.
</Notification>
```

### Error

Used for error messages. Text is bolded to show importance.

```
<Notification variant="error">
  Looks like our registration system is temporarily down. You’ll need to come back another time to register for My Account. In the meantime, return to TELUS.com.
</Notification>
```
