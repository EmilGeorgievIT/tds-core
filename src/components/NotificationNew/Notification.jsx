import React from 'react'
import PropTypes from 'prop-types'

import safeRest from '../../safeRest'

import Container from '../Grid/Container/Container'
import Icon from '../Icon/Icon'
import ColoredTextProvider from '../Typography/ColoredTextProvider/ColoredTextProvider'
import Paragraph from '../Typography/Paragraph/Paragraph'

import styles from './Notification.modules.scss'

const iconByVariant = {
  success: 'checkmark',
  error: 'exclamation-point-circle'
}

const isImportant = variant => variant === 'success' || variant === 'error'

const renderIcon = glyph => (
  <span className={styles.icon}>
    <Icon glyph={glyph} aria-hidden="true" />
  </span>
)

const renderContent = (variant, children) => {
  const content = (
    <Paragraph bold={isImportant(variant)}>
      {children}
    </Paragraph>
  )

  if (variant === 'error') {
    return (
      <ColoredTextProvider colorClassName={styles.errorText}>
        {content}
      </ColoredTextProvider>
    )
  }

  return content
}

/**
 * A banner that highlights important messages.
 */
const Notification = ({ variant, children, ...rest }) => (
  <div {...safeRest(rest)} className={styles[variant]}>
    <Container limitWidth>
      <div className={styles.flexRow}>
        {isImportant(variant) ? renderIcon(iconByVariant[variant]) : undefined}

        {renderContent(variant, children)}
      </div>
    </Container>
  </div>
)

Notification.propTypes = {
  /**
   * The appearance.
   */
  variant: PropTypes.oneOf([
    'instructional',
    'branded',
    'success',
    'error'
  ]),
  /**
   * The message. Can be raw text or text components.
   */
  children: PropTypes.node.isRequired
}

Notification.defaultProps = {
  variant: 'instructional'
}

export default Notification
