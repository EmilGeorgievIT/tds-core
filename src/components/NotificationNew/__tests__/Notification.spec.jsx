import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Icon from '../../Icon/Icon'

import Notification from '../Notification'

describe('<Notification />', () => {
  const doShallow = ({ ...overrides }) => (
    shallow(<Notification {...overrides}>Some content</Notification>)
  )

  it('renders', () => {
    const notification = doShallow()

    expect(toJson(notification)).toMatchSnapshot()
  })

  it('can have a variant', () => {
    let notification = doShallow({ variant: undefined })
    expect(notification).toHaveClassName('instructional')

    notification = doShallow({ variant: 'success' })
    expect(notification).toHaveClassName('success')
  })

  it('adds an icon only to the error and success variants', () => {
    let notification = doShallow({ variant: 'error' })
    expect(notification.find('NotificationIcon').dive()).toContainReact(<Icon glyph="exclamation-point-circle" aria-hidden="true" />)

    notification = doShallow({ variant: 'success' })
    expect(notification.find('NotificationIcon').dive()).toContainReact(<Icon glyph="checkmark" aria-hidden="true" />)

    notification = doShallow({ variant: 'branded' })
    expect(notification.find('NotificationIcon').dive()).toHaveText('')
  })

  it('passes additional HTML attributes to the containing element', () => {
    const notification = doShallow({ id: 'hello', title: 'my title' })

    expect(notification).toHaveProp('id', 'hello')
    expect(notification).toHaveProp('title', 'my title')
  })

  it('does not allow custom CSS', () => {
    const notification = doShallow({ className: 'my-custom-class', style: { color: 'hotpink' } })

    expect(notification).not.toHaveProp('className', 'my-custom-class')
    expect(notification).not.toHaveProp('style')
  })
})
