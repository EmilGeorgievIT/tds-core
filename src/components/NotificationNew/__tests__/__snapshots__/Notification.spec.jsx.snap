// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notification /> renders 1`] = `
<div
  className="instructional"
>
  <Container
    limitWidth={true}
  >
    <div
      className="content"
    >
      <NotificationIcon
        variant="instructional"
      />
      <p
        className="text"
      >
        Some content
      </p>
    </div>
  </Container>
</div>
`;
